blog/03-04-2023-how-to-write-a-multi-signatures-contract-on-celo-using-hardhat-part-1-2/how-to-write-a-multi-signatures-contract-on-celo-using-hardhat-part-1-2.md

```diff
@@ -0,0 +1,653 @@
+
+Let's explain what our `execTransaction` function does. First, let's take a look at the arguments it takes:
+
+- `address to`: The address of the target contract or user.
+- `uint256 value`: The value to be sent with this transaction in Celo. If you're calling a smart contract and the function doesn't require Celo, pass in 0.
+- `bytes memory data`: The call data of the transaction. You can pass in some call data like any transaction. If you're calling a smart contract function, this is where you'll enter the encoded function signature and arguments.
+- `uint256 txnGas`: The maximum gas to be used by this transaction. If the call consumes more gas than this, the transaction will fail.
+- `bytes memory signatures`: The signatures of all the owners that approved this transaction.
+
+Inside the logic of the function, our first verification is to check that `signatures` argument is at least as long as the length of a regular signature multiplied by the threshold of the contract. We use the `hashTypedDataV4` function in `EIP712.sol` from OpenZeppelin to independently hash all the arguments, the same way the owners would have hashed them prior to signing them, with the following code: `hashTypedDataV4(keccak256(abi.encode(_TRANSACTION_TYPEHASH, to, value, keccak256(data), txnGas, _nonce)))`. This way, we can retrieve the signer of each signature to verify that they are the right owners.
+
+We use a loop to verify each signature for up to the amount of threshold using an internal function, `_getCurrentOwner()`. This internal function uses assembly, a lower-level programming language that allows us to perform simple operations (in this case, byte manipulation) at lower gas cost. The main purpose of this internal function is to split the different signatures that we had previously concatenated all together and finally call `ecrecover()` to retrieve the signer of each signature.
+
+We then verify that this owner has not already signed this transaction and that they are an owner. After verifying all signatures, we save in storage that the owner has signed this transaction, and we increment the nonce so that it cannot be used again. This prevents replay attacks.
+
+We then execute the call (the transaction) and record the gas used by looking at `gasLeft()` before and after, making sure it did not consume more than the limit passed as an argument. Finally, we emit an event with the result of the call, either `TransactionExecuted` or `TransactionFailed`, with the details of this transaction.
+
+## Compile the contract
+
+We have finished writing our contract, and now it's time to see if our contract compiles! To do so, let's run:
+
+```bash
+npx hardhat compile
+```
+
+Like this:
+
+![NPX Hardhat Compile](./images/npx_hardhat_compile.png)
+
+## Clone or fork this project repository
+
+The full contract is available at [Celo Multi-Signatures Hardhat](https://github.com/marc-aurele-besner/celo-multi-signatures-hardhat)
```

Comment:
Please add Conclusion and About Me Section At the end You can Add Details about part 1 in conclusion .